descriptors: Use xpub at last hardened step if possible

achow101 opened the pull request:
When a descriptor contains a private key and has hardened derivation steps, derive the xpub at the last hardened step and convert the derivation into origin info. So the result is that a `BIP32PubkeyProvider` containing a `xprv` with hardened derivation steps and ending with unhardened derivation becomes a `OriginPubkeyProvider` with those hardened derivation steps as the origin info and a new `BIP32PubkeyProvider` containing the `xpub` and the unhardened derivation steps.

For example:
`wpkh(xprv..../0'/0'/*)` becomes `wpkh([d34db33f/0'/0']xpub.../*)`

If the descriptor has only hardened steps, it is not modified.

This change allows descriptor wallets to derive from such descriptors without needing to be unlocked.

achow101 committed:
descriptors: Use xpub at last hardened step if possible

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -713,6 +716,51 @@ std::unique_ptr<PubkeyProvider> ParsePubkeyInner(const Span<const char>& sp, boo
+            std::copy(extpubkeyid.begin(), extpubkeyid.begin() + 4, last_hardened_info.fingerprint);
+
+            KeyPath::iterator it = path.end() - 1;
+            for (; it >= path.begin(); it--) {
```
Travis doesn't like this: https://travis-ci.org/bitcoin/bitcoin/jobs/651243106#L4064

```
Error: attempt to decrement a dereferenceable (start-of-sequence) iterator.
```

achow101 replied:
I was able to use `std::find_if` for this.

Sjors reviewed (commented):
Concept ACK

instagibbs commented:
Is there a practical consideration for this change? It seems to make sense, just wondering if downstream applications are made easier by this or something.

achow101 commented:
> Is there a practical consideration for this change?

It would allow for descriptor wallets that import a descriptor with a private key and hardened steps to be able to derive scriptPubKeys without needing to be unlocked.

instagibbs commented:
Adding that to OP would be OP

concept ACK!

Sjors reviewed (commented):
0e54ba78ea7e7468ce5a185c4195b0dd58f3f82d looks good, expect the duplicate line.

That said, I feel a bit ambivalent about adding origin info to descriptors that don't already have it. For example `pkh(xprv9...C2U/0'/0'/0)` becomes `pkh([bd16bee5/0'/0']xpub6AVD...pQKh/0)`. Perhaps normalisation should only happen for descriptors with origin info. Also, an xpub encodes its depth, so if in the example the original xpriv depth wasn't 0 it would be weird (or maybe not; depends on if we interpret the fingerprint as depth 0).

I personally haven't seen real world descriptors without origin info yet, so I'm sure how they should be handled.

Also, what would be the next step to have `InferDescriptor` benefit from this change? It would be nice to have `getaddressinfo` in return an xpub, instead of individual public key, for BIP44/49/84 style watch-only (legacy) wallets.

achow101 commented:
> That said, I feel a bit ambivalent about adding origin info to descriptors that don't already have it. For example `pkh(xprv9...C2U/0'/0'/0)` becomes `pkh([bd16bee5/0'/0']xpub6AVD...pQKh/0)`. Perhaps normalisation should only happen for descriptors with origin info.

Then we wouldn't achieve the primary goal of this PR. People could import xprvs without origin info that have some hardened derivation steps and we wouldn't be able to derive the public keys from them.

> Also, an xpub encodes its depth, so if in the example the original xpriv depth wasn't 0 it would be weird (or maybe not; depends on if we interpret the fingerprint as depth 0).

The xpub at the end would still contain that depth information. Even without an origin info, an xprv could have a non-zero depth but that doesn't bother us.

> Also, what would be the next step to have `InferDescriptor` benefit from this change? It would be nice to have `getaddressinfo` in return an xpub, instead of individual public key, for BIP44/49/84 style watch-only (legacy) wallets.

I think that requires changes to `SigningProvider`.

achow101 committed:
Add DerivePath to CExtKey

Utility function to derive at a path from an CExtKey instead
of implementing this externally every time we want to derive at
a path.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -713,6 +721,48 @@ std::unique_ptr<PubkeyProvider> ParsePubkeyInner(const Span<const char>& sp, boo
+
+            // Only continue with this last hardened stuff if there are actually hardened steps in path
+            if (!last_hardened_info.path.empty()) {
+                // Derive that last hardened key
```
Done

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -713,6 +721,48 @@ std::unique_ptr<PubkeyProvider> ParsePubkeyInner(const Span<const char>& sp, boo
+
+                // Shorten path to the unhardened stuff
+                if (last_hardened_info.path.size() < path.size()) {
+                    // At least on step is unhardened
```
Fixed

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -713,6 +721,48 @@ std::unique_ptr<PubkeyProvider> ParsePubkeyInner(const Span<const char>& sp, boo
+                }
+
+                // Make a BIP32PubkeyProvider for the unhardened derivation
+                extpubkey = extkey.Neuter();
```
Removed

Sjors commented:
ACK b674297c7c491ea09cef171ed083533f23c9ffa9

I've never imported keys without origin info, but if you find it useful, I'm fine with supporting it.

achow101 commented:
Had to rebase

Sjors commented:
Code review re-ACK f8274751f6eb7dbd2207ff4232131dfc0f9ff978 (rebased on #18034)

sipa commented:
I discussed this with @achow101 IRL.

One alternative I'd like to see explored is instead of rewriting the descriptor itself, having an "xpub cache" added to the descriptor interface (and later the wallet itself), similar to the existing pubkey cache, but shared across indexes. This would have all the availability benefits of a change like this, but also speed up operations on descriptors that include several non-hardened steps in xpubs.

Sjors commented:
That would take case of this TODO, right?

https://github.com/bitcoin/bitcoin/blob/ab9de435880c9d77e4137b65050591ef2d14f809/src/script/descriptor.cpp#L286-L291

Rewriting the descriptor has merit in any case; a descriptor like `wpkh([d34db33f/xpub /0'/0'].../*)` makes no sense, because you can't actually derive any addresses without also knowing the `xpriv`. Whereas with `wpkh([d34db33f/0'/0']xpub.../*)` you can. Normalising makes them more straight-forward to share.

For BIP44/49/84 style descriptors (`wpkh([d34db33f/84'/0'/0']xpub.../0/*)`) it would indeed make sense to cache an xpub for the `0` after the visible `xpub`. But I think that can be hidden as an ephemeral implementation detail; I don't think it needs to be serialised into the wallet payload. Unlike the individual key cache, the savings don't scale with wallet size.

instagibbs commented:
something "like" this is probably nice to have for multisig setups. I'm looking into it for a `dumpdescriptors` RPC command for that type of setup
